spring-batch-s3/src/main/java/org/springframework/batch/extensions/s3/stream/S3MultipartOutputStream.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2006-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.batch.extensions.s3.stream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.s3.S3Client;
+
+/**
+ * An {@link OutputStream} that writes data to an S3 object using multipart upload. It
+ * uses a {@link PipedInputStream} and a {@link PipedOutputStream} to allow writing data
+ * asynchronously while uploading it in parts. This stream is suitable for large file
+ * uploads.
+ *
+ * @author Andrea Cioni
+ */
+public class S3MultipartOutputStream extends OutputStream {
+
+	private static final Logger logger = LoggerFactory.getLogger(S3MultipartOutputStream.class);
+
+	private final PipedInputStream pipedInputStream;
+
+	private final PipedOutputStream pipedOutputStream;
+
+	private final ExecutorService singleThreadExecutor;
+
+	private volatile boolean uploading;
+
+	private final S3Uploader multipartUpload;
+
+	public S3MultipartOutputStream(S3Client s3Client, String bucketName, String key) throws IOException {
+		this(new S3MultipartUploader(s3Client, bucketName, key));
+	}
+
+	public S3MultipartOutputStream(S3Uploader s3Uploader) throws IOException {
+		this.pipedInputStream = new PipedInputStream();
+		this.pipedOutputStream = new PipedOutputStream(this.pipedInputStream);
+		this.singleThreadExecutor = Executors.newSingleThreadExecutor();
+		this.uploading = false;
+		this.multipartUpload = s3Uploader;
+	}
+
+	@Override
+	public void write(int b) throws IOException {
+		if (!this.uploading) {
+			this.uploading = true;
+
+			startUpload();
+		}
+		this.pipedOutputStream.write(b);
+	}
+
+	private void startUpload() {
+		this.singleThreadExecutor.execute(() -> {
```

Comment:
You mentioned that currently writer is synchronous/blocking. But here you offload an upload operation into a new thread even without thread pooling .
Why?